### PR TITLE
feat(parser): add select statement

### PR DESCRIPTION
**Pull request by andylokandy**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

- Implement select statements with trivial joins.
- Refresh AST and parser for type name
- Add various DDL statements

## Changelog

- New Feature

## Related Issues

- https://github.com/datafuselabs/databend/issues/4916
- https://github.com/datafuselabs/databend/issues/1218

### CODE REVIEW

File: common/ast/src/parser/rule/statement.rs [hunk 1]
```diff
@@ -47,3 +49,210 @@ pub fn statement(i: Input) -> IResult<Statement> {
...
+                table,
+                alias: alias.map(|(_, alias)| alias),
+            }
+        },
+    )(i)
+}
+
+pub fn table_reference(i: Input) -> IResult<TableReference> {
+    rule! (
+        #parenthesized_joined_tables
```
Comment (leiysky) [inline common/ast/src/parser/rule/statement.rs:171]: A `table_reference` could also be `parenthesized_subquery`.

File: common/ast/src/parser/rule/statement.rs [hunk 2]
```diff
@@ -47,3 +49,210 @@ pub fn statement(i: Input) -> IResult<Statement> {
...
+        rule! {
+            USING ~ "(" ~ #comma_separated_list1(ident) ~ ")"
+        },
+        |(_, _, idents, _)| JoinCondition::Using(idents),
+    );
+    let join_condition = rule! { #join_condition_on | #join_condition_using };
+
+    let join = map(
+        rule! {
+            #aliased_table ~ #join_operator? ~ JOIN ~ #cut(table_reference) ~ #cut(join_condition)
```
Comment (leiysky) [inline common/ast/src/parser/rule/statement.rs:207]: Maybe the `#aliased_table` should be a `table_reference`?

**Comment by leiysky** [comment]
Rest LGTM.

File: common/ast/src/parser/rule/statement.rs (hunk 2 above)
Comment (andylokandy) [reply common/ast/src/parser/rule/statement.rs]: addressed

File: common/ast/src/parser/rule/statement.rs (hunk 1 above)
Comment (andylokandy) [reply common/ast/src/parser/rule/statement.rs]: addressed

**Comment by BohuTANG** [comment]
'make lint' locally will make the check CI happy :)

**Comment by andylokandy** [comment]
> 'make lint' locally will make the check CI happy :)

But it all-checks on my laptop

**Comment by BohuTANG** [comment]
@mergify update

**Comment by BohuTANG** [comment]
CI failures:
https://github.com/datafuselabs/databend/runs/6088060806?check_suite_focus=true#step:3:962

**Comment by BohuTANG** [comment]
Same error on my local.

**Comment by sundy-li** [comment]
Maybe some files are not commited.

**Comment by andylokandy** [comment]
I've force-pushed again, let's wait and see.

File: common/ast/src/parser/ast/expression.rs [hunk 3]
```diff
@@ -100,22 +100,20 @@ pub enum Expr {
...
+    Float,
+    Double,
     Date,
-    Time,
+    DateTime,
     Timestamp,
-    Interval,
-    Text,
+    Varchar,
+    Array,
```
Comment (sundy-li) [inline common/ast/src/parser/ast/expression.rs:114]: We decide to have `Array(TypeName)`. It's more generic.

Comment (andylokandy) [reply common/ast/src/parser/ast/expression.rs:114]: What is the grammar like for the type generic?

Comment (leiysky) [reply common/ast/src/parser/ast/expression.rs:114]: That's great. By the way, shall we have some structural types like https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#struct_type

Comment (andylokandy) [reply common/ast/src/parser/ast/expression.rs:114]: Seem postgresql chose a [different syntax](https://www.postgresql.org/docs/9.1/arrays.html#:~:text=CREATE%20TABLE%20tictactoe%20(%0A%20%20%20%20squares%20%20%20integer%5B3%5D%5B3%5D%0A)%3B)

**Comment by andylokandy** [comment]
> Same error on my local.

Confirmed after re-pulling. Seems like the git has silently merged upstream changes without modifying the local file.

File: common/ast/src/parser/ast/expression.rs (hunk 3 above)
Comment (sundy-li) [reply common/ast/src/parser/ast/expression.rs:114]: Currently, structural type is replaced by Variant...

Comment (sundy-li) [reply common/ast/src/parser/ast/expression.rs:114]: > What is the grammar like for the type generic?

```
Array(tinyint)
Array(Varchar)
Array(int unsigned)
```

**Comment by BohuTANG** [comment]
Ci failure:
image

CI error log is at:
https://github.com/datafuselabs/databend/runs/6089539318?check_suite_focus=true#step:3:965

**Comment by BohuTANG** [comment]
BTW, if a CI is failure, we can check the detail logs:
image

File: common/ast/src/parser/ast/expression.rs (hunk 3 above)
Comment (andylokandy) [reply common/ast/src/parser/ast/expression.rs]: addressed

**Review by sundy-li — APPROVED** [review]

**Pull request merged**